File: Source/ObjC/SWGradientNode.m
//
//  SWGradientNode.m
//  SWGradientNode
//
//  Created by NSSimpleApps on 05.06.16.
//  Copyright © 2016 NSSimpleApps. All rights reserved.
//

#import "SWGradientNode.h"
#import "SWGradientShader.h"

#if TARGET_OS_IPHONE

#define SWPointMake CGPointMake

#define SWRectGetMidX CGRectGetMidX
#define SWRectGetMidY CGRectGetMidY

#elif TARGET_OS_MAC

#define SWPointMake NSMakePoint

#define SWRectGetMidX NSMidX
#define SWRectGetMidY NSMidY

#endif

static NSString *const SWUniformColor = @"uniformColor";
static NSString *const SWUniformLocation = @"uniformLocation";

@interface SWGradientNode ()

@property (strong) SKUniform *uniformCenter;

@property (strong) SKUniform *uniformStartAngle;

@property (strong) SKUniform *uniformInnerRadius;

@end

@implementation SWGradientNode

- (instancetype)initWithColors:(NSArray<SKColor *> *)colors
                     locations:(NSArray<NSNumber *> *)locations
                        bounds:(SWRect)bounds
                  gradientType:(SWGradientType)gradientType {
    
    self = [super initWithTexture:nil
                            color:[SKColor clearColor]
                             size:bounds.size];
    
    if (self) {
        
        self.uniformCenter = [SKUniform uniformWithName:@"uniformCenter"
                                           floatVector2:GLKVector2Make(0.5, 0.5)];
        self.uniformStartAngle = [SKUniform uniformWithName:@"uniformStartAngle" float: 0.0];
        self.uniformInnerRadius = [SKUniform uniformWithName:@"uniformInnerRadius" float: 0.0];
        
        self.position = SWPointMake(SWRectGetMidX(bounds), SWRectGetMidY(bounds));
        
        if (colors.count == 1) {
            
            self.color = colors[0];
            
        } else {
            
            NSMutableArray<SKUniform *> *uniforms = [NSMutableArray arrayWithObjects:self.uniformCenter, self.uniformStartAngle, self.uniformInnerRadius, nil];
            
            [colors enumerateObjectsUsingBlock:^(SKColor * _Nonnull color, NSUInteger idx, BOOL * _Nonnull stop) {
                
                CGFloat red = 0, green = 0, blue = 0, alpha = 0;
                
                [color getRed:&red green:&green blue:&blue alpha:&alpha];
                
                GLKVector4 vector4 = GLKVector4Make(red, green, blue, alpha);
                
                SKUniform *colorUniform = [SKUniform uniformWithName:[NSString stringWithFormat:@"%@%@", SWUniformColor, @(idx)]
                              floatVector4:vector4];
                
                [uniforms addObject:colorUniform];
            }];
            
            if (locations != nil && ((locations.count - 1) == colors.count)) {
                
                [locations enumerateObjectsUsingBlock:^(NSNumber * _Nonnull location, NSUInteger idx, BOOL * _Nonnull stop) {
                    
                    NSString *name = [NSString stringWithFormat:@"%@%@", SWUniformLocation, @(idx + 1)];
                    
                    SKUniform *locationUniform = [SKUniform uniformWithName:name
                                                                      float:location.floatValue];
                    [uniforms addObject:locationUniform];
                }];
                
            } else {
                
                for (NSUInteger elem = 1; elem < colors.count; elem++) {
                    
                    NSString *name = [NSString stringWithFormat:@"%@%@", SWUniformLocation, @(elem)];
                    
                    CGFloat location = (float)elem/colors.count;
                    
                    SKUniform *locationUniform = [SKUniform uniformWithName:name
                                                                      float:location];
                    [uniforms addObject:locationUniform];
                }
            }
            
            SWGradientShader *gradientShader = nil;
            
            if (gradientType == SWGradientTypeSmooth) {
                
                gradientShader =
                [[SWSmoothGradientShader alloc ] initWithColor:SWUniformColor
                                                      location:SWUniformLocation
                                                        center:self.uniformCenter.name
                                                    startAngle:self.uniformStartAngle.name
                 innerRadius:self.uniformInnerRadius.name];
                
            } else if (gradientType == SWGradientTypeStep) {
                
                gradientShader =
                [[SWStepGradientShader alloc ] initWithColor:SWUniformColor
                                                    location:SWUniformLocation
                                                      center:self.uniformCenter.name
                                                  startAngle:self.uniformStartAngle.name
                 innerRadius:self.uniformInnerRadius.name];
            }
            
            self.shader = [[SKShader alloc] initWithSource:[gradientShader shaderWithNumberOfColors:colors.count - 1]
                                                  uniforms:uniforms];
        }
    }
    return self;
}

- (SWPoint)center {
    
    GLKVector2 floatVector = self.uniformCenter.floatVector2Value;
    
    CGFloat x = floatVector.x;
    
    #if TARGET_OS_IPHONE
    
        CGFloat y = 1 - MIN(MAX(floatVector.y, 0.0), 1.0);
    
    #elif TARGET_OS_MAC
    
        CGFloat y = MIN(MAX(floatVector.y, 0.0), 1.0);
    #endif

    return SWPointMake(x, y);
}

- (void)setCenter:(SWPoint)center {
    
    CGFloat x = MIN(MAX(center.x, 0.0), 1.0);
    
    #if TARGET_OS_IPHONE
    
    CGFloat y = 1 - MIN(MAX(center.y, 0.0), 1.0);
    
    #elif TARGET_OS_MAC
    
    CGFloat y = MIN(MAX(center.y, 0.0), 1.0);
    
    #endif
    
    self.uniformCenter.floatVector2Value = GLKVector2Make(x, y);
}

- (CGFloat)startAngle {
    
    return self.uniformStartAngle.floatValue;
}

- (void)setStartAngle:(CGFloat)startAngle {
    
    self.uniformStartAngle.floatValue = fmodf(startAngle, 2 * M_PI) / (2 * M_PI);
}


- (CGFloat)innerRadius {
    
    return self.uniformInnerRadius.floatValue;
}

- (void)setInnerRadius:(CGFloat)innerRadius {
    
    self.uniformInnerRadius.floatValue = MIN(MAX(innerRadius, 0.0), 0.5);
}

@end

File: Source/Swift/SWGradientNode.swift
//
//  SweepGradientNode.swift
//  SweepGradient
//
//  Created by NSSimpleApps on 22.05.16.
//  Copyright © 2016 NSSimpleApps. All rights reserved.
//

import SpriteKit

#if os(iOS)
    public typealias SWPoint = CGPoint
    public typealias SWRect = CGRect
    
#elseif os(OSX)
    public typealias SWPoint = NSPoint
    public typealias SWRect = NSRect
#endif

private let SWUniformColor = "uniformColor"
private let SWUniformLocation = "uniformLocation"

/// smooth and step types of gradient
public enum SWGradientType {
    
    case smooth
    case step
}

/// smooth and step gradient node
open class SWGradientNode: SKSpriteNode {
    
    required public init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    fileprivate let uniformCenter = SKUniform(name: "uniformCenter", float: GLKVector2Make(0.5, 0.5))
    
    /// The center of sweeping gradient between (0.0, 0.0) and (1.0, 1.0).
    /// (0.0, 0.0) is a bottom left corner (OSX) or top left corner (iOS).
    /// (1.0, 1.0) is a top right (OSX) or bottom left corner (iOS).
    /// Default is (0.5, 0.5)
    open var center: SWPoint {
        
        get {
            
            let floatVector = self.uniformCenter.floatVector2Value
            
            let x = CGFloat(floatVector.x)
            
            #if os(iOS)
                let y = 1 - CGFloat(min(max(floatVector.y, 0.0), 1.0))
            #elseif os(OSX)
                let y = CGFloat(min(max(floatVector.y, 0.0), 1.0))
            #endif
            
            return SWPoint(x: x, y: y)
        }
        
        set {
            
            let x = Float(min(max(newValue.x, 0.0), 1.0))
            
            #if os(iOS)
                let y = 1 - Float(min(max(newValue.y, 0.0), 1.0))
            #elseif os(OSX)
                let y = Float(min(max(newValue.y, 0.0), 1.0))
            #endif
            
            self.uniformCenter.floatVector2Value = GLKVector2Make(x, y)
        }
    }
    
    fileprivate let uniformStartAngle = SKUniform(name: "uniformStartAngle", float: 0.0)
    
    /// The first color of the gradient starts at this angle in radians between 0 and 2*PI
    /// 0 is to the right along the x axis.
    /// All colors are located in counter-clockwise order.
    /// Default is 0
    open var startAngle: Float {
        
        get {
            
            return self.uniformStartAngle.floatValue
        }
        
        set {
            
            self.uniformStartAngle.floatValue = fmodf(newValue, 2 * Float.pi) / (2 * Float.pi)
        }
    }
    
    
    fileprivate let uniformInnerRadius = SKUniform(name: "uniformInnerRadius", float: 0.0)
    
    /// The inner radius of the gradient
    /// Node will not draw any gradient inside circle with this radius
    /// Value between 0 and 0.5
    /// Default is 0
    open var innerRadius: Float {
        
        get {
            
            return self.uniformInnerRadius.floatValue
        }
        
        set {
            
            self.uniformInnerRadius.floatValue = Float(min(max(newValue, 0.0), 0.5))
        }
    }
    
    public init(colors: [SKColor], locations: [Float]?, bounds: SWRect, gradientType: SWGradientType) {
        
        super.init(texture: nil, color: SKColor.clear, size: bounds.size)
        
        self.position = SWPoint(x: bounds.midX, y: bounds.midY)
        
        if colors.count == 1 {
            
            self.color = colors[0]
            
        } else {
            
            var uniforms: [SKUniform] = [self.uniformCenter, self.uniformStartAngle, self.uniformInnerRadius]
            
            for (index, color) in colors.enumerated() {
                
                var red = CGFloat(0.0), green = CGFloat(0.0), blue = CGFloat(0.0), alpha = CGFloat(0.0)
                
                color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
            
                let vector4 = GLKVector4Make(Float(red), Float(green), Float(blue), Float(alpha))
                let colorUniform = SKUniform(name: "\(SWUniformColor)\(index)", float: vector4)
                uniforms.append(colorUniform)
            }
            
            if let locations = locations , locations.count - 1 == colors.count {
                
                for (index, location) in locations.enumerated() {
                    
                    let name = "\(SWUniformLocation)\(index + 1)"
                    
                    let locationUniform = SKUniform(name: name, float: location)
                    uniforms.append(locationUniform)
                }
                
            } else {
                
                (1..<colors.count).forEach({ (elem: Int) in
                    
                    let name = "\(SWUniformLocation)\(elem)"
                    
                    let locationUniform = SKUniform(name: name, float: Float(elem)/Float(colors.count))
                    uniforms.append(locationUniform)
                })
            }
            
            let gradientShader: SWGradientShader
            
            switch gradientType {
            
            case .smooth:
                gradientShader =
                SWSmoothGradientShader(uColor: SWUniformColor,
                                       uLocation: SWUniformLocation,
                                       uCenter: self.uniformCenter.name,
                                       uStartAngle: self.uniformStartAngle.name, uInnerRadius: self.uniformInnerRadius.name)
                
            case .step:
                gradientShader =
                    SWStepGradientShader(uColor: SWUniformColor,
                                         uLocation: SWUniformLocation,
                                         uCenter: self.uniformCenter.name,
                                         uStartAngle: self.uniformStartAngle.name, uInnerRadius: self.uniformInnerRadius.name)
            }
            
            self.shader = SKShader(source: gradientShader.shaderWithNumberOfColors(colors.count - 1))
            self.shader?.uniforms = uniforms
        }
    }
}
